### ISSUE TITLE
listenbrainz signon fails

### ISSUE THREAD

**Issue by pgueckel** [mariusor/mpris-scrobbler#67]
See:
https://bugzilla.redhat.com/show_bug.cgi?id=1746472
In short, I am unable to get the scrobbler to create a file, read signon information from a file manually created ($HOME/.local/share/mpris-scrobbler/credentials) etc.

**Comment by mariusor** [comment]
Hi @pgueckel thank you for your interest in my project.

I think your issue is a duplicate of #63 . 

Can you wait for @jwflory to release v0.3.5 and use that, or install it from source, or  do the same with latest master and check if the problem is fixed?

**Comment by pgueckel** [comment]
I prefer to wait for the release. While I know how to install from source (tarballs), I don't care to install unsigned software and that doesn't come from the fedora and associated repos. Thanks.

So, I should communicate in future under #63 instead of here?

**Comment by jwflory** [comment]
I will submit a package update this weekend.

**Comment by mariusor** [comment]
@pgueckel you can try the workarounds I proposed in #63 . I think setting the environment variable `XDG_DATA_HOME` to `~/.local/share` would work.

**Comment by pgueckel** [comment]
I tested the value of XDG_DATA_HOME before I reported the problem. It's value is default on Fedora. If I set the value, would that not taint the result of the test of the upcoming new release? How would we know if the release fixed the problem, when we apply workarounds before testing it?

**Comment by pgueckel** [comment]
Well, I couldn't wait :-) I can, after all, simply unset the variable once the new version is released.

I played a file with vlc. I logged in to listenbrainz and it says I have 0 listens :-( Then I tried smplayer. Still 0 listens.

I don't even know if mpris-scrobbler.service is running. When I run systemctl|grep mpris it shows nothing. How can I get it to run right now? I tried systemctl --user start --now mpris-scrobbler.service and --user enable and --force and whatever, but grep scrobbler shows nothing. How do I even know it's running?

**Comment by mariusor** [comment]
To see if the scrobbler works you can just execute it from the command line after you compile it without worrying the service:

```$ ./build/mpris-scrobbler -vv```

For the service to be accessible you need to actually install the scrobbler using `sudo make install` after you compiled it (to remove them do `sudo make uninstall`). 

This copies the service file and the binaries to their right place so, the systemd commands work. It should all be detailed in the manual install section of the [readme](https://github.com/mariusor/mpris-scrobbler/blob/master/README.md).

Also, out of the box vlc and smplayer might not have support for the MPRIS interface to report that tracks change, so you probably should use a more traditional music player: audacious, rhythmbox, amarok, clementine, or even proprietary ones like the Spotify player.

To check if vlc and smplayer have MPRIS enabled you can play something in them and then in a terminal run this command:

```sh
dbus-send --print-reply --type=method_call --dest=org.freedesktop.DBus / org.freedesktop.DBus.ListNames | grep mpris

 "string" "org.mpris.MediaPlayer2.vlc"
```

**Comment by pgueckel** [comment]
I'd really like to use amarok, my standard player, but they haven't, after over a year, managed to get an rpm built for KDE5 (I don't want to taint my system with any KDE4 stuff). Rhythmbox is installed, but it wants access to all of my files and I had troubles with it years ago: I think it tries to put the music files into ITS hierarchy. I just don't like it! I don't know if Spotify is in the Fedora/RPMFusion repos and it is a pay system and I don't have an account. I am playing my own music, over 1000 ripped CDs, plus stuff from NaxosMusicLibrary and YouTube. I will see about audadious and clementine. I have never had any draw to them, but perhaps they might be usable :-(

OK, I'll give that command a try. VLC would be my top choice, given that amarok is still not out.

**Comment by pgueckel** [comment]
I ran the command and I get the "string" output with vlc, so vlc apparently does work with mpris. When I run mpris-scrobbler -vv I get a lot of output regarding title, artist, album etc, but there is still nothing in listenbrainz.

**Comment by mariusor** [comment]
Without that output I can't help, sorry. 

Please run `mpris-scrobbler -vvv` and upload the logs to a pastebin.

**Comment by pgueckel** [comment]
OK, I will do, but presently I have noticed something else :-(

There is a confounding issue with graphical shells: they do not read environment variables. I finally managed to find a solution for KDE; it is to stop using .bash_profile, but rather .profile. I noticed that when I set XDG_DATA_HOME in a terminal window, it is set only in that teminal window and not set in other terminal windows. It seems to me that this could have a bearing on whether the variable value is being read by programs that need to have XDG_DTA_HOME set. So, I have just edited my .profile and I will a execute system reboot. Wait a moment and I will se if anything changes ;-)

**Comment by pgueckel** [comment]
Confirmed: with export XDG_DATA_HOME=~/.local/share in .profile, the variable value is now known in all terminal windows. Also, the mpris-scrobbler.service daemon is running all of the time, even after asystem reboot, without me having to restart it. Also, the credentials file is correctly installed with the token stored and the enabled. So, I presume that the only action I need to take, after logging in, is to open a terminal, type mpris-scrobbler-signon listenbrainz. Correct? Do I signon like this every time I close the terminal window and reopen it? Every time I log out of KDE and log back in? Or is this, like the running daemon, also active permanently? And how about when I open VLC: do I need to rerun the mpris signon each time I open and close VLC, even if I am still logged in to the same KDE session? Is there a harm in running the signon program when I am already signed on, or is this like trying to run a second instance?

OK, assuming you have given me understandable answers ;-) back to your request, that I run mpris-scrobbler -vvv. Since it is already running all of the time (I configured it using the systemctl command that is given in the docs), would this work or is this like trying to run a second instance of the daemon? Do I need to stop the running daemon to execute the -vvv command? how to I stop it without screwing up the prefectly installed systemd setup I have?

**Comment by mariusor** [comment]
`mpris-scrobbler-signon listenbrainz` is needed only once, after that the informatoin is saved to `~/.local/share/mpris-scrobbler/credentials` 

It's the same with the other services.

And you can't have multiple scrobblers instances running at the same time, if the systemd service is running, if you execute it in the command line, it will exit without doing anything.

To test in the command line, you need to stop the service `systemctl --user stop mpris-scrobbler` and only after execute the scrobbler.

**Comment by pgueckel** [comment]
Hey! Something weird just happened. Maybe mpris-scrobbler.service isn't correctly installed by systemctl? Or maybe it is? Anyway, I will tell you what I just did.

I am still logged in to the same KDE session. I still have the same terminal window open, but I was going to call it a day and I closed VLC. Then, I thought: what the heck, I'll risk it and just run the mpris-scrobbler -vvv command anyway, even though the daemon is running. I reopened VLC and typed  the -vvv command into the still open terminal and a whole lot of output appeared, the final line catching my eye:

INFO     api::submitted_to[listenbrainz.org:0]: ok

I logged in to listenbrainz and lo and behold, I now have a first listen logged!

Now, I must say that this is all very complicated: all of this starting and typing commands and whatnot. Is there a way for me just to log in to my computer, start playing music with VLC or amarok and just have my listens recorded without me having to do all of the signon each and every time?

**Comment by mariusor** [comment]
As I said earlier, once you "signed-on" that's enough. If the mpris-scrobbler systemd service is enabled and it starts fine when you log-in, you don't have to do anything.

**Comment by pgueckel** [comment]
OK. I will see how this goes for me and get back to you in a couple of days, once I have had some time to play with it. Thanks :-)

**Comment by mariusor** [comment]
I'll keep my fingers crossed for you.
Keep me posted if you have other issues, but for now I I'll close this issue, as you seem to be sending tracks to listenbrainz.

**Comment by pgueckel** [comment]
There is still something wrong. It will not scrobble when I run mpris-scrobbler as installed to systemd as per instructions. The only way it scrobbles is when started from the command line (with the systemd script stopped), and then it only does one track and fails, with the unable to load scrobble warning.

See pastebin.com/gMzQdxWw

What I did was stop the non-functioning instance with the stop command, as you instructed systemctl stop etc. Then, I opened a terminal and ran mpris-scrobbler and opened vlc to play a track. It showed in listenbrainz, but when the track ended, it did not register the end and it showed as playing in listenbrainz for a long time, before finally disappearing, even thogh it had played to the very end. Then, still running the instance of mpris-scrobbler, i opened smplayer and played a track, which is shown in the pastebin post, but nothing ever appeared in listenbrainz. Finally, still with mpris-scrobbler running, I opened vlc again and playd another track, but nothing happened in listenbrainz.

The daemon can only scrobble one track and then it fails to "load scrobble" and sometimes it doesn't even detect the end of a track, so that play is lost entirely, even though it was shown as playing for a while on listenbrainz.

What is happening here?

**Comment by pgueckel** [comment]
See here: https://pastebin.com/ygeJyu2U

I listened to a very long work with vlc and, as you see, the instance of mpris-scrobbler started from the command line kept updating to listenbrainz. Approx. every 10 minutes, the playing now and X minutes ago lines would be updated, but when the track ended, it failed to send the info to listenbrainz and the listen is lost.

**Comment by mariusor** [comment]
I see multiple problems in your messages, and I am sorry you're having such a terrible experience with mpris-scrobbler.

1. Installed mpris-scrobbler doesn't work, command line mpris-scrobbler works: you probably have installed the old version, so the systemd user service points to that. 

2. After the last track in a playlist finishes, the scrobbler keeps pushing now-playing events to listenbrainz < Valid Issue that I'll open a ticket for.

3. No scrobbles from smplayer < From your log I can deduce that smplayer sends multiple properties changed signals when the track change and the scrobbler can't correlated them into a valid message to send to listenbrainz. This is a known issue, that I currently have no solution for.

4. With vlc the scrobbler sends now listening events but not scrobble events < probably the same issue as above.  

All in all, I have to say that I don't have a better solution for you than to advise you to change players to something with a more compatible MPRIS implementation.

**Comment by pgueckel** [comment]
vlc is a major player that I also use on my android phone, but I just discovered, last night after I wrote you, JuK! :-) It is also mpris-compatible and works very well, but it is dropping some tracks, too. I didn't run it with -vvv, but I did, of course, use the command line start. This is the player I will use mostly, on the laptop, for the forseeable future, for music listening.

I rarely use smplayer, so that is not a big issue. I just wanted to test how other players work.

I have mpris-scrobbler-0.3.5-1.fc30.x86_64 installed, since it was made available on the Fedora 30 updates-testing repo. I cannot have both the old and new versions installed, since dnf erases the old one when upgrading to the new one. How can I unpoint the systemd user service and repoint it correctly? Should I uninstall and reinstall mpris-scrobbler or is there a way to remove the systemd link(s)?

Yes, I think I will switch to Juk, as it is kde5/qt5 and a full-featured native player (vlc is qt5, too, I believe, but an independent, albeit excellent, project). I think we are making great progress :-)

Have you opened a ticket for #2 above? I want to watch that one.

**Comment by jwflory** [comment]
> The daemon can only scrobble one track and then it fails to "load scrobble" and sometimes it doesn't even detect the end of a track, so that play is lost entirely, even though it was shown as playing for a while on listenbrainz.

To double-check, is the service running under your individual user ID and not `root`? i.e. as your logged-in user, `$ systemctl --user start mpris-scrobbler`

> How can I unpoint the systemd user service and repoint it correctly? Should I uninstall and reinstall mpris-scrobbler or is there a way to remove the systemd link(s)?

Restarting the service (`systemctl --user restart mpris-scrobbler.service`) should be sufficient. The Fedora package uses the generated systemd service file shown [here](https://github.com/mariusor/mpris-scrobbler/blob/master/units/systemd-user.service.in). So long as the process is fully stopped, restarting should be a seamless upgrade.

**Comment by pgueckel** [comment]
I have tried all of this: make sure that there is no mpris-scrobbler running as root, only --user; stopped the service and restarted the service; verified that it is running and that the player is able to send to listenbrainz, but...

It does not do anything! The only way I can get anything sent to listenbrainz is when I start mpris-scrobbler from the command line—and that, as described above, only sends the first track played.
